### PR TITLE
API: warn if stairs used in way that is likely not desired

### LINKED ISSUE #26752: [Bug]: `ax.stairs()` creates inaccurate `fill` for the plot
### Bug summary

I was working on a code to achieve the following look: 

image

I got there by combining `ax.plot()` and `ax.fill_between()` functions. But then I've learnt about the `ax.stairs()` function which is suppose to plot in a step-like fashion and allows to add fill under the line **in one go**.

However, when I tried `ax.stairs()` I got the following result:

image

The bug (I think):
* the fill is completely wrong - it doesn't stretch between `x-axis` and the line

Secondary issues:
* I don't think there's a way to have a separate colour for line and fill at the moment, when I set a colour, line disappears - this is especially visibile when I set e.g. `alpha=0.1`
* I've got a feeling that somehow the grid ends up on top of the plot when I call `ax.stairs()`

### Code for reproduction

```python
python
import matplotlib.pyplot as plt

x = [1, 4, 10, 12, 15, 20]
y = [2, 4, 1, 3, 1, 4]

fig, axs = plt.subplots(
    2, 1,
    figsize=(16,10),
    facecolor="white",
    sharex=True,
    sharey=True
)
fig.suptitle(
    "ax.stairs()",
    fontsize=20,
)

# plot points
for ax in axs:
    ax.scatter(
        x, y,
        marker="X",
        s=275,
        linewidth=1.5,
        edgecolor="black",
        facecolor="white",
        zorder=100  # crude way to keep the points always on top
    )

axs[0].stairs(
    values=y,  # step height
    edges=[x[0], *x],  # edge width
    baseline=None,
    color="red",
    fill=True,
    label="steps-pre: duplicate 1st element in `edges=[x[0], *x]`"
)
axs[1].stairs(
    values=y,  # step height
    edges=[*x, x[-1]],  # edge
    baseline=None,
    color="green",
    fill=True,
    alpha=0.1,
    label="steps-post: duplicate last element in `edges=[*x, x[-1]]`"
)

# set axis limits
axs[0].set_ylim(0, 5.5)
axs[0].set_yticks(range(1, 6))
axs[0].set_xlim(0, 21)
axs[1].set_xticks(range(1, 21))

# add legend
for ax in axs:
    ax.legend(
        fontsize=16,
        loc="upper center",
        facecolor="whitesmoke",
        edgecolor="none"
    )

    # Update font size of the 'x' and 'y' axis labels
    ax.tick_params(
        axis="both",
        labelsize=16
    )

    ax.grid(
        color="gainsboro",
        linewidth=0.5
    )

# Remove spines
for position in ["top", "right"]:
    for ax in axs:
        ax.spines[position].set_visible(False)

plt.tight_layout()
```
```


### Actual outcome

![image](https://github.com/matplotlib/matplotlib/assets/8880777/f726f74e-5d28-488d-bf66-6536093b3227)


### Expected outcome

![image](https://github.com/matplotlib/matplotlib/assets/8880777/88a1c5e4-f50f-4b30-8f41-59f69aeee169)


### Additional information

_No response_

### Operating system

Windows 

### Matplotlib Version

3.7.2

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

3.10.2

### Jupyter version

IPython          : 8.14.0 ipykernel        : 6.25.1 ipywidgets       : not installed jupyter_client   : 8.3.0 jupyter_core     : 5.3.1 jupyter_server   : not installed jupyterlab       : not installed nbclient         : not installed nbconvert        : not installed nbformat         : not installed notebook         : not installed qtconsole        : not installed traitlets        : 5.9.0

### Installation

pip

### PR DESCRIPTION
**Pull request by tacaswell**
Closes #26752

I think the best we can do here is warn the user they are about to do something silly.  Raising on this combination seems too harsh.

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -7194,9 +7194,16 @@ def stairs(self, values, edges=None, *,
...
             path is drawn.
 
+            If None, then drawn as an unclosed Path.
+
         fill : bool, default: False
             Whether the area under the step curve should be filled.
 
+            Passing both ``fill=True` and ``baseline=None`` will likely result in
+            undesired filling as the first and last points will be connected
+            with a straight line with the fill between and the stairs.
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:7204]:
```suggestion
            with a straight line with the fill between the line and the stairs.
```

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -7234,6 +7241,15 @@ def stairs(self, values, edges=None, *,
                                    fill=fill,
                                    **kwargs)
         self.add_patch(patch)
+        if baseline is None and fill:
+            _api.warn_external(
+                f"Both {baseline=} and {fill=} have been passed. "
+                "Because baseline in None, the Path used to draw the stairs will "
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:7247]:
```suggestion
                "baseline=None is only intended for unfilled stair plots. "
                "Because baseline in None, the Path used to draw the stairs will "
```

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -2342,6 +2342,23 @@ def test_hist_zorder(histtype, zorder):
...
+        "Because baseline in None, the Path used to draw the stairs will "
+    )
+    with pytest.warns(UserWarning, match=warn_match):
+        ax.stairs(
+            [4, 5, 1, 0, 2],
+            [1, 2, 3, 4, 5, 6],
+            facecolor="blue",
+            baseline=baseline,
+            fill=fill
+        )
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_axes.py:2359]: Why not inline?
```suggestion
    with pytest.warns(UserWarning, match="baseline=None and fill=True"):
        ax.stairs(
            [4, 5, 1, 0, 2],
            [1, 2, 3, 4, 5, 6],
            facecolor="blue",
            baseline=None,
            fill=True
        )
```

Comment (tacaswell) [reply lib/matplotlib/tests/test_axes.py:2359]: a first version of this test was trying to format and test for the whole message including the f-strings but for reasons I don't understand it was not matching so I went with just the middle part of the message with dropped the f-strings.

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -7194,9 +7194,16 @@ def stairs(self, values, edges=None, *,
...
             path is drawn.
 
+            If None, then drawn as an unclosed Path.
+
         fill : bool, default: False
             Whether the area under the step curve should be filled.
 
+            Passing both ``fill=True` and ``baseline=None`` will likely result in
+            undesired filling as the first and last points will be connected
+            with a straight line with the fill between the line and the stairs.
```
Comment (rcomer) [inline lib/matplotlib/axes/_axes.py:7204]:
```suggestion
            undesired filling: the first and last points will be connected
            with a straight line and the fill will be between this line and the stairs.
```
I found this sentence a little hard to parse.

**Review by rcomer — APPROVED** [review]
I have a small wording suggestion.  Please take or leave and merge when ready.

**Comment by rcomer** [comment]
Appveyor failure is everywhere so merging anyway.

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -7234,6 +7241,16 @@ def stairs(self, values, edges=None, *,
                                    fill=fill,
                                    **kwargs)
         self.add_patch(patch)
+        if baseline is None and fill:
+            _api.warn_external(
+                f"Both {baseline=} and {fill=} have been passed. "
+                "baseline=None is only intended for unfilled stair plots. "
+                "Because baseline is None, the Path used to draw the stairs will "
+                "not be closed, thus because fill is True the polygon will be closed "
+                "by drawing an (unstroked) edge from the first to last point.  It is "
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py:7250]: I find this a bit difficult to parse, as there are two 'because' in the sentence, and also two conjunctions (thus because) in the second half.
